File: Day #112/solution.go
// Day 112: LCA with parent pointers - equalize depths, walk up together. O(h).
package main

import "fmt"

type Node struct {
	Val            int
	Parent, L, R *Node
}

func mk(p *Node, v int) *Node { return &Node{Val: v, Parent: p} }

func depth(n *Node) int {
	d := 0
	for n != nil {
		n = n.Parent
		d++
	}
	return d
}

func lca(a, b *Node) *Node {
	da, db := depth(a), depth(b)
	for da > db {
		a = a.Parent
		da--
	}
	for db > da {
		b = b.Parent
		db--
	}
	for a != b {
		a = a.Parent
		b = b.Parent
	}
	return a
}

func main() {
	root := &Node{Val: 3}
	root.L = mk(root, 5)
	root.R = mk(root, 1)
	root.L.L = mk(root.L, 6)
	root.L.R = mk(root.L, 2)
	root.R.L = mk(root.R, 0)
	root.R.R = mk(root.R, 8)
	root.L.R.L = mk(root.L.R, 7)
	root.L.R.R = mk(root.L.R, 4)
	fmt.Println(lca(root.L, root.R).Val)         // 3
	fmt.Println(lca(root.L.R.L, root.L.R.R).Val) // 2
}

File: Day #112/solution.py
# Day 112: LCA with parent pointers - equalize depths, walk up together. O(h).
class Node:
    def __init__(self, val, parent=None):
        self.val = val
        self.parent = parent
        self.l = None
        self.r = None


def depth(n):
    d = 0
    while n:
        n = n.parent
        d += 1
    return d


def lca(a, b):
    da, db = depth(a), depth(b)
    while da > db:
        a, da = a.parent, da - 1
    while db > da:
        b, db = b.parent, db - 1
    while a is not b:
        a, b = a.parent, b.parent
    return a


if __name__ == "__main__":
    root = Node(3)
    root.l = Node(5, root); root.r = Node(1, root)
    root.l.l = Node(6, root.l); root.l.r = Node(2, root.l)
    root.r.l = Node(0, root.r); root.r.r = Node(8, root.r)
    root.l.r.l = Node(7, root.l.r); root.l.r.r = Node(4, root.l.r)
    print(lca(root.l, root.r).val)            # 3
    print(lca(root.l.r.l, root.l.r.r).val)    # 2
